sigs/jvm/rfcs/20190904-java-tensors.md

```diff
@@ -0,0 +1,301 @@
+We should not introduce more complexity unless needed. Plus, it can be generally safe
+to assume that when dealing with constants, we can map implicitely a given Java type to a common TF data type
+(e.g. a `int` value generates a `TInt32` constant, a `float` value generates a `TFloat` constant, etc.), while allowing
+the user to force a different data type if needed.
+
+The underlying implementation though must be changed to avoid using reflective techniques to generate the tensor and
+must be more explicit. To simplify this task, we should limit the possible factories to rank-0 or rank-1
+constants (which satisfies most of the cases) and another to pass explicitly a `Tensor` instance for constant of a 
+higher rank. For example:
+
+```java
+public final class Constant<U> extends PrimitiveOp {
+
+    public static final Constant<TInt32> create(Scope scope, Integer value) {
+        ...
+    }
+    
+    public static final Constant<TInt32> create(Scope scope, Integer... values) {
+        ...
+    }
+    
+    public static final <U extends Tensor<?>> Constant<U> create(Scope scope, U data) {
+        ...
+    }
+}
+
+Constant<TInt32> scalar = tf.constant(10);  // single argument constructor is chosen
+
+Constant<TInt32> vector = tf.constant(10, 11, 42);  // variadic constructor is chosen
+
+Constant<TInt32> matrix = tf.constant(TInt32.ofShape(2, 2).row(10, 11).row(30, 40).done());
+
+```
+
+## Addendum
+
+### NdArray Iterative Data Initialization
+
+In some previous examples, some tensors are initialize using a new method called `row`. This method is intented
+to be added to the original [`NdArray` API](https://github.com/karllessard/community/blob/master/sigs/jvm/rfcs/20190606-java-tensor-io.md#ndarray-api)
+submitted in the previous RFC, with the objective of allowing the allocation and initialization
+of a small tensor in a single line to mimics standard multidimensional Java arrays.
+
+The idea is to initialize the data of each row of the `NdArray` sequentially by invoking `row(...)`, 
+and then invoke the `done()` method to return the newly allocated `NdArray` instance. A "row" in this terminology
+means a vector on the last axis (dimension) of an array.
+
+Let's compare standard Java arrays with those new methods as they are used with tensors:
+```java
+
+// Create a 2D matrix
+
+int[][] matrix2d_s = new int[][] { {1, 2}, {3, 4} };
+
+TInt32 matrix2d_t = TInt32.ofShape(2, 2).row(1, 2).row(3, 4).done();
+
+// Create a 3D matrix
+
+int[][][] matrix3d_s = new int[][][] { { {1, 2}, {3, 4} }, { {5, 6}, {7, 8} }, { {9, 10}, {11, 12} } };
+
+TInt32 matrix3d_t = TInt32.ofShape(3, 2, 2)
+  .row(1, 2).row(3, 4)
+  .row(5, 6).row(7, 8)
+  .row(9, 10).row(11, 12)
+  .done();
+```
+Of course, the standard Java arrays gives a better view on where is located a given row, by embedding it
+through a series of brackets. The actual proposal is focusing on simplicity and performance more that readability.
+*Please, feel free to suggest any better solution.*
```

Comment:
I'm not too fond of this implicit row wise construction as I got completely lost trying to figure out where I was in the Tensor at any given call, plus it means the TensorBuilder is carrying around implicit state saying where it wants to write to next. 

At the expense of readability I think it would be preferable to use something like
```
TInt32.ofShape(3,2,2)
.set(0,0,new NDArray<>(1,2))
.set(0,1,new NDArray<>(3,4))
.set(1,0,new NDArray<>(5,6))
.set(1,1,new NDArray<>(7,8))
.set(2,0,new NDArray<>(9,10))
.set(2,1,new NDArray<>(11,12));
```
Replacing `new NDArray<>()` with whatever the appropriate constructor is. To make the varargs work we might have to permute the order of arguments (i.e. `new NDArray<>(),int..`). It could also be a TensorBuilder and require a `done()` method to construct too, I've no strong feelings either way.

Reply:
hehe, yes me neither to be honest, I just drop the basic idea here to let others suggest something better :)

The example you gave is pretty much already supported by the `NdArray` API, except that the coordinates are provided after the value (because its a vararg that can conflict with an int scalar value). For example, you can set a scalar value like this:
```java
TInt32 t  = ...;
t.set(100, 0, 0).set(200, 1, 0);
```
We could add something similar but vectors, which will accept a vector, something like
TInt32.ofShape(3, 2, 2)
    .set(vector(1, 2), 0, 0)
    .set(vector(3, 4), 0, 1)
    .set(vector(5, 6), 1, 0)
    ...
````
where `vector` instantiate a rank-1 array... or simply a primitive or auto-boxed array. What do you think? I understand that specifying the coordinates for each vector is more explicit but more verbose at the same time. I wouldn't mind though, personally.

Reply:
Yeah, I think that's better. I'm in favour of explicitness when constructing values.